### PR TITLE
AF_XDP socket extension using libxdp api

### PR DESCRIPTION
**Pull request by plangarbalint**
The aim of this change is to extend existing packet sending methods with AF_XDP socket using libxdp api.

By using AF_XDP socket, the time needed for packet sending can be decreased substantially.

- all the relevant changes are inside a #ifdef HAVE_LIBXDP guard 
- --batch-size flag is introduced to tcpreplay interface. AF_XDP socket is used typically with batch of packets, the default value is 25 however the user can set it using --batch-size flag, the maximum value is set to 4096 now (see my reasoning at the very end of this comment). 
- Due to batch processing feature tcpr_sleep will be called only in case --batch-size=1, since it doesn't make sense to sleep the relative time between two packets when batch-size is higher than 1. 
- Due to batch processing feature statistics calculated during packet processing are also updated (e.g. number of packets, sum of bytes sent etc.)
- Due to the hierarchical nature of packet sending selection (if ... else if ... include guards in sendpacket.c), --enable-force-libxdp is introduced which can be used at the configuration step to force libxdp packet sending method
- The dependencies of libxdp is checked at the configuration step and the output table of configuration step is also updated with libxdp information 
- The size of UMEM memory is set to nb_of_frames = 4096   * frame_size = 4096   as a result the current implementation assumes that a packet size is never larger than 4096 byte. The new implementation makes it possible to reuse memory, since packet sending is organized into batches. Each of the batches requires a fixed amount of memory, after each of the batches the program checks whether all the packets in the given batch have been sent, if yes, the program can start working on the next batch.
- This implementation is also compatible with the existing tcpreplay interface e.g. -K (to cache packets), -topspeed, --loop flags are still available

Potential improvement suggestions which are not part of this PR:
- Optimize reserved memory. The minimum reserved memory that is required equals to batch_size * frame_size, since user should copy the packet data to an umem frame as a first step of packet sending. As a result it would be an advantage to let users decide how much memory should be reserved by setting batch_size and frame_size on the interface. Batch size is introduced to the interface however it is not connected to umem creation. As mentioned before this implementation reserves 4096 * 4096 bytes, however it should be customized. In case batch-size is set to 25 then only the first 25 * 4096 bytes of memory is used.

### CODE REVIEW

**Comment by fklassen** [comment]
Thanks for the PR. I have wanted to build and AF_XDP implementation for years. Targetting version 4.5 which hopefully will happen soon.

File: configure.ac [hunk 1]
```diff
@@ -826,7 +830,13 @@ fi
 
 # libpcap can require libnl
 AC_SEARCH_LIBS([nl_handle_alloc], [nl],
-        [AC_MSG_NOTICE([Unable to find nl library - may be needed by libpcap])])
+        [AC_MSG_NOTICE([Unable to find xdp library - may be needed by libpcap])])
```
Comment (fklassen) [inline configure.ac:833]: is this a copy/paste error?

Comment (SPYFF) [reply configure.ac:833]: Yes, its very likely copy-paste mistake/quick experiment left in the final PR.
This line should be unmodified, as in the original code.

**Comment by fklassen** [comment]
Bringing into a development branch to resolve conficts and test, while maintaining history.

File: src/defines.h.in [hunk 2]
```diff
@@ -44,6 +44,28 @@
 
 #include "tcpr.h"
 
+#ifdef HAVE_BPF
+#include <net/bpf.h>
+#define PCAP_DONT_INCLUDE_PCAP_BPF_H 1
+#endif
+
```
Comment (fklassen) [inline src/defines.h.in:51]: copy/paste error

**Pull request merged**